### PR TITLE
Add support for Delve debugger

### PR DESCRIPTION
**Pull request by amrfaissal**
This adds support for debugging tool Delve. The command `dlv` starts a debugging REPL that allows the user to debug his application using Delve tool.
image

### CODE REVIEW

**Comment by sergeylanzman** [comment]
@amrfaissal How I can check it?

**Comment by astaxie** [comment]
```
cmd/commands/dlv/dlv.go:98:2: should merge variable declaration with assignment on next line (S1021)
```

**Comment by amrfaissal** [comment]
To learn more this is the documentation: https://github.com/derekparker/delve/tree/master/Documentation
I found also a video on Youtube: https://www.youtube.com/watch?v=zgLjVD5ZSOc

I will label this PR as work-in-progress as I am planning to do some improvements.

**Comment by astaxie** [comment]
@amrfaissal @sergeylanzman when this PR is done, I would like to release a new version: `1.8.1`

**Comment by amrfaissal** [comment]
I think this PR is ready for review/merge. Let me know what you think.

**Comment by astaxie** [comment]
AFAIK the dlv has different installation for different os. I see dlv dependence in vendor. Does the user still need to install dlv to use this command?

File: cmd/commands/dlv/dlv.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	defer os.Remove(fp)
+
+	abs, err := filepath.Abs(debugname)
+	if err != nil {
+		beeLogger.Log.Fatalf("%v", err)
+	}
+
+	//
+	// Create and start the debugger server
+	//
```
Comment (astaxie) [inline cmd/commands/dlv/dlv.go]: you can remove these empty comments

File: cmd/commands/dlv/dlv.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+		WorkingDir:  "./",
+		ProcessArgs: []string{abs},
+	}, false)
+	if err := server.Run(); err != nil {
+		beeLogger.Log.Fatalf("Could not start debugger server: %v", err)
+	}
+
+	//
+	// Start the Delve client REPL
+	//
```
Comment (astaxie) [inline cmd/commands/dlv/dlv.go]: you can remove these empty comments

File: utils/utils.go [hunk 3]
```diff
@@ -332,3 +333,80 @@ func PrintErrorAndExit(message, errorTemplate string) {
...
+	allargs := []string{command}
+	allargs = append(allargs, args...)
+	goBuild := exec.Command("go", allargs...)
+	goBuild.Stderr = os.Stderr
+	return goBuild.Run()
+}
+
+// GoBuild runs the "go build" command on the specified package
+func GoBuild(debugname, pkg string) error {
+	buildFlags := "-ldflags='-linkmode internal'"
```
Comment (astaxie) [inline utils/utils.go]: could you add the comments here why need to use `-ldflags='-linkmode internal'` is this for dlv only? As this is the util package, so maybe other package will use it

**Comment by amrfaissal** [comment]
This is the purpose of my improvements: No need to install dlv command line. Delve can be used either using the binary or the API. In the last commits, I am using the API by starting the server and client terminal. The server used in `bee` supports `MultiClient` mode so the user can use `bee dlv` to use it as client and open another terminal and start a different client using: `dlv connect [addr]:[port]`.

File: cmd/commands/dlv/dlv.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	debugname := "debug"
+	addr := fmt.Sprintf("127.0.0.1:%d", port)
+	return runDelve(addr, debugname)
+}
+
+// runDelve runs the Delve debugger server
+func runDelve(addr, debugname string) int {
+	beeLogger.Log.Info("Starting Delve Debugger...")
+
+	err := utils.GoBuild(debugname, packageName)
```
Comment (astaxie) [inline cmd/commands/dlv/dlv.go]: from your screenshot you don't need the packageName for it. So `bee dlv` would start to work. But it looks you pass this to the `go build -o`. If it's not set then the output would be empty. I think you need to add a deafult value for it.

**Comment by astaxie** [comment]
> The server used in bee supports MultiClient mode so the user can use bee dlv to use it as client and open another terminal and start a different client using: dlv connect [addr]:[port].

Thanks for the detail introduction. So we need to update our readme and docs

**Comment by amrfaissal** [comment]
I'll update the README as well before the release of `1.8.1`.

File: cmd/commands/dlv/dlv.go (hunk 4 above)
Comment (amrfaissal) [inline cmd/commands/dlv/dlv.go]: No  `debugname` is passed to `go build -o` then we specify what package to build. If no package is supplied `go build` builds the current directory. So no need to add a default value.

File: utils/utils.go (hunk 3 above)
Comment (amrfaissal) [inline utils/utils.go]: Yes you're right. Other packages might use different `ldflags`, but here I am just making sure Go's internal linker is used. I will move this function to `dlv.go`.

**Pull request merged**